// @dendronhq/engine-server/lib/testUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rmTmpDendronDir = exports.createFileStorage = exports.SCHEMA_FOO_PARENT_ID = exports.FOO_SCHEMA = exports.FOO_NOTE = exports.NOTE_FOO_CHILD_ONE_ID = exports.NOTE_FOO_PATH = exports.ROOT_ID = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const store_1 = __importDefault(require("./drivers/file/store"));
exports.ROOT_ID = "root";
exports.NOTE_FOO_PATH = "foo";
exports.NOTE_FOO_CHILD_ONE_ID = "foo.one";
exports.FOO_NOTE = {
    id: "foo",
    title: "foo",
};
exports.FOO_SCHEMA = {
    title: "foo",
    id: "b111db5b-bc52-4977-893b-307522f89ea3",
    fname: "foo.schema",
    child: {
        id: "foo.two",
    },
};
exports.SCHEMA_FOO_PARENT_ID = "root.schema";
function createFileStorage(root) {
    // TODO: update
    return new store_1.default({
        root,
    });
}
exports.createFileStorage = createFileStorage;
function rmTmpDendronDir(root) {
    return fs_extra_1.default.removeSync(root);
}
exports.rmTmpDendronDir = rmTmpDendronDir;
