// @dendronhq/engine-server/lib/topics/markdown/plugins/dendronLinksPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dendronLinksPlugin = void 0;
const common_server_1 = require("@dendronhq/common-server");
const LINK_REGEX = /^\[\[(.+?)\]\]/;
function locator(value, fromIndex) {
    return value.indexOf("[", fromIndex);
}
function dendronLinksPlugin(opts = {}) {
    const permalinks = opts.permalinks || [];
    const defaultPageResolver = (name) => [
        name.replace(/ /g, "").toLowerCase(),
    ];
    const pageResolver = opts.pageResolver || defaultPageResolver;
    const newClassName = opts.newClassName || "new";
    const wikiLinkClassName = opts.wikiLinkClassName || "internal";
    const defaultHrefTemplate = (permalink) => `#/page/${permalink}`;
    const hrefTemplate = opts.hrefTemplate || defaultHrefTemplate;
    const aliasDivider = opts.aliasDivider || "|";
    function isAlias(pageTitle) {
        return pageTitle.indexOf(aliasDivider) !== -1;
    }
    function parseAliasLink(pageTitle) {
        const [displayName, name] = pageTitle.split(aliasDivider);
        return { name, displayName };
    }
    function parsePageTitle(pageTitle) {
        if (isAlias(pageTitle)) {
            return parseAliasLink(pageTitle);
        }
        return {
            name: pageTitle,
            displayName: pageTitle,
        };
    }
    function inlineTokenizer(eat, value) {
        const match = LINK_REGEX.exec(value);
        if (match) {
            const pageName = match[1].trim();
            const { name, displayName } = parsePageTitle(pageName);
            const pagePermalinks = pageResolver(name);
            let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
            const exists = permalink !== undefined;
            if (!exists) {
                permalink = pagePermalinks[0];
            }
            let classNames = wikiLinkClassName;
            if (!exists) {
                classNames += " " + newClassName;
            }
            // normalize
            if (permalink) {
                permalink = common_server_1.removeMDExtension(permalink);
            }
            return eat(match[0])({
                type: "wikiLink",
                value: name,
                data: {
                    alias: displayName,
                    permalink,
                    exists,
                    hName: "a",
                    hProperties: {
                        className: classNames,
                        href: hrefTemplate(permalink),
                    },
                    hChildren: [
                        {
                            type: "text",
                            value: displayName,
                        },
                    ],
                },
            });
        }
        return;
    }
    inlineTokenizer.locator = locator;
    // @ts-ignore
    const Parser = this.Parser;
    // @ts-ignore
    const inlineTokenizers = Parser.prototype.inlineTokenizers;
    // @ts-ignore
    const inlineMethods = Parser.prototype.inlineMethods;
    inlineTokenizers.wikiLink = inlineTokenizer;
    inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");
    // Stringify for wiki link
    // @ts-ignore
    const Compiler = this.Compiler;
    if (Compiler != null) {
        const visitors = Compiler.prototype.visitors;
        if (visitors) {
            // eslint-disable-next-line func-names
            visitors.wikiLink = function (node) {
                const data = node.data;
                if (!node || !node.data || !node.data.alias) {
                    throw Error("no alias found");
                }
                if (data.useId) {
                    if (!data.note) {
                        throw Error("no note");
                    }
                    node.value = data.note.id;
                }
                if (data.toMd) {
                    return `[${data.alias}](${data.prefix || ""}${node.value})`;
                }
                if (node.data.alias !== node.value) {
                    return `[[${node.data.alias}${aliasDivider}${node.value}]]`;
                }
                return `[[${node.value}]]`;
            };
        }
    }
}
exports.dendronLinksPlugin = dendronLinksPlugin;
