// chrome-debugging-client/dist/lib/create-http-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = require("http");
const event_promise_1 = require("./event-promise");
function createHTTPClient(host, port) {
    return new HTTPClient(host, port);
}
exports.default = createHTTPClient;
class HTTPClient {
    constructor(host, port) {
        this.host = host;
        this.port = port;
    }
    get(path) {
        return __awaiter(this, void 0, void 0, function* () {
            const { host, port } = this;
            const request = http_1.get({ host, port, path });
            const response = yield getResponse(request);
            const statusCode = response.statusCode;
            const body = yield readResponseBody(response);
            if (typeof statusCode === "number" && statusCode !== 200) {
                throw new ResponseError(body, statusCode);
            }
            return body;
        });
    }
}
function getResponse(request) {
    return __awaiter(this, void 0, void 0, function* () {
        return event_promise_1.eventPromise(request, "response", "error");
    });
}
function readResponseBody(response) {
    return __awaiter(this, void 0, void 0, function* () {
        let body = "";
        response.setEncoding("utf8");
        response.on("data", chunk => {
            body += chunk;
        });
        yield event_promise_1.eventPromise(response, "end", "error");
        return body;
    });
}
/* tslint:disable:max-classes-per-file */
class ResponseError extends Error {
    constructor(message, statusCode) {
        super(message);
        this.statusCode = statusCode;
    }
}
