// @dendronhq/common-server/lib/logger.js
"use strict";
// import pino from "pino";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logAndThrow = exports.pino = exports.createLogger = exports.Logger = void 0;
const common_all_1 = require("@dendronhq/common-all");
const pino_1 = __importDefault(require("pino"));
exports.pino = pino_1.default;
class Logger {
    constructor(opts) {
        this.debug = (msg) => {
            this._log(msg);
        };
        this.info = (msg) => {
            this._log(msg);
        };
        this.error = (msg) => {
            this._log(msg);
        };
        this.name = opts.name;
        this.level = opts.level;
    }
    _log(msg) {
        let ctx = "";
        if (msg.ctx) {
            ctx = msg.ctx;
        }
        // eslint-disable-next-line no-console
        console.log(this.name, ctx, msg);
    }
}
exports.Logger = Logger;
function createLogger(name, dest) {
    const level = common_all_1.env("LOG_LEVEL", { shouldThrow: false }) || "error";
    const nameClean = name || common_all_1.env("LOG_NAME");
    // if (logDst === "stdout") {
    //   // TODO: tmp disable pino logging on stdout
    //   const out = pino({ name: nameClean, level });
    //   return out;
    // } else {
    if (dest) {
        return pino_1.default(pino_1.default.destination(dest)).child({ name: nameClean, level });
    }
    else {
        // const logDst = env("LOG_DST", { shouldThrow: false })
        // if (logDst) {
        //   return pino(pino.destination(logDst)).child({ name: nameClean, level });
        // }
        return pino_1.default({ name: nameClean, level });
    }
}
exports.createLogger = createLogger;
function logAndThrow(logger, msg) {
    logger.error(msg);
    throw JSON.stringify(msg);
}
exports.logAndThrow = logAndThrow;
